### DISCUSSION TITLE
Target types enum tests: scope enum constants in pattern matching expressions when the left han operator is an enum

### DISCUSSION THREAD

**Discussion by henning-moe** [dotnet/csharplang#8903] [Language Ideas] [answered]
When doing pattern matching with enum values it could save some code if we didn't need to repeat the enum type in the expression since that's already known to the compiler.

Take the code for the current version of the language:
```csharp
enum E
{ 
  A = 1;
  B = 2;
  C = 3;
}

public string M(E value) => value is E.A or E.C ? "A" : "C";
```

What would perhaps be more pleasant is if the enum constants were scoped to the right hand expression if the left hand expression is an enum.

```csharp
public string M(E value) => value is A or C ? "Eff" : "Oof";
```

The reason I came over this is because I'm working with gRPC and the enum checks becomes quite a mouthful.
Here's code in question from an API I'm making that shows the issue:

```csharp
if (request.NewState is VesselTrackingStateRequest.Types.VesselTrackingStatus.Undefined)
{
    // ...
}
```

Maybe this could just use `Undefined` since the compiler knows that it has to be a member for that enum anyway:

```csharp
if (request.NewState is Undefined)
{
    // ...
}
```

## Questions ##

Should this be allowed in normal equality operators?

```csharp
if(myEnum == A) { }
```

Maybe it could more generally be a target type expression, and not just for comparisons?

```csharp
void M1(E value) { /* ... */ }
void M2()
{
  E myValue = A;
  M1(A);
}
```

Or it could just allow static members of a type that is of that same type in order to allow things like this:

```sealed record E(int value)
{
  public static E A { get; } = new E(0);
  public static E B { get; } = new E(1);
  public static E C { get; } = new E(0);
}

void M(E value)
{
  if(value == A) {} // translates to if(value == E.A)
}
```

**Comment by huoyaoyuan** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/8641 https://github.com/dotnet/csharplang/discussions/4479

**Reply by henning-moe** [reply]
How did I miss those :/
